File: 0079.WordSearch.go
func exist(board [][]byte, word string) bool {
    if len(board) == 0 || len(board[0]) == 0 {
        return false
    }
    m, n := len(board), len(board[0])
    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            if dfs(board, i, j, word) {
                return true
            }
        }
    }
    return false
}

func dfs(board [][]byte, i, j int, word string) bool {
    if len(word) == 0 {
        return true
    }
    if i < 0 || i >= len(board) || j < 0 || j >= len(board[0]) || board[i][j] != word[0] {
        return false
    }
    board[i][j] ^= 1<<7
    found := dfs(board, i-1, j, word[1:]) || dfs(board, i+1, j, word[1:]) || dfs(board, i, j-1, word[1:]) || dfs(board, i, j+1, word[1:])
    board[i][j] ^= 1<<7
    return found
}

File: 0079.WordSearch.py
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        if not word:
            return True
        if not board:
            return False
        for i in xrange(len(board)):
            for j in xrange(len(board[0])):
                if self.dfs(board, word, i, j):
                    return True
        return False

    def dfs(self, board, word, i, j):
        if not word:
            return True
        if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]) or board[i][j] != word[0]:
            return False
        temp = board[i][j]
        board[i][j] = '#'
        ans = self.dfs(board, word[1:], i-1, j) or self.dfs(board, word[1:], i+1, j) or self.dfs(board, word[1:], i, j-1) or self.dfs(board, word[1:], i, j+1)
        board[i][j] = temp
        return ans
